std-bits/base/src/main/java/org/enso/base/parser/NumberWithSeparators.java

```diff
@@ -0,0 +1,393 @@
+      case '.' -> Optional.of(NO_DOT);
+      case ',' -> Optional.of(NO_COMMA);
+      default -> Optional.empty();
+    };
+  }
+
+  private static Optional<NumberWithSeparators> matchForUnknown(char decimal) {
+    return switch (decimal) {
+      case Constants.UNKNOWN -> Optional.of(UNKNOWN);
+      case '.' -> Optional.of(UNKNOWN_DOT);
+      case ',' -> Optional.of(UNKNOWN_COMMA);
+      default -> Optional.empty();
+    };
+  }
+
+  private static Optional<NumberWithSeparators> matchForSpace(char decimal) {
+    return switch (decimal) {
+      case Constants.UNKNOWN -> Optional.of(SPACE_UNKNOWN);
+      case '.' -> Optional.of(SPACE_DOT);
+      case ',' -> Optional.of(SPACE_COMMA);
+      default -> Optional.empty();
+    };
+  }
+
+  private static Optional<NumberWithSeparators> matchForSwiss(char decimal) {
+    return switch (decimal) {
+      case Constants.UNKNOWN -> Optional.of(SWISS_UNKNOWN);
+      case '.' -> Optional.of(SWISS_DOT);
+      case ',' -> Optional.of(SWISS_COMMA);
+      default -> Optional.empty();
+    };
+  }
+
+  static class Constants {
+    static final char NONE = '\0';
+    static final char UNKNOWN = '\uFFFD';
+  }
+
+  static boolean isDigit(char c) {
+    return (c >= '0' && c <= '9');
+  }
+
+  private final char thousands;
+  private final char decimal;
+
+  NumberWithSeparators(char thousands, char decimal) {
+    this.thousands = thousands;
+    this.decimal = decimal;
+  }
+
+  public char getThousands() {
+    return thousands;
+  }
+
+  public char getDecimal() {
+    return decimal;
+  }
+
+  NumberParseResult parse(CharSequence value, int idx, boolean integer) {
+    var separators = Separators.parse(value, idx, integer);
+    if (separators == null) {
+      return new NumberParseFailure("Invalid separators.");
+    }
+
+    if (thousands != Constants.UNKNOWN && (integer || decimal != Constants.UNKNOWN)) {
+      // If we have a fixed format then we can parse the number.
+      return integer
+          ? parseFixedInteger(value, idx, separators.endIdx(), separators.first())
+          : parseFixed(value, idx, separators.endIdx(), separators.first(), separators.second());
+    }
+
+    return integer
+        ? parseUnknownInteger(
+            value, idx, separators.endIdx(), separators.first(), separators.count())
+        : parseUnknown(
+            value,
+            idx,
+            separators.endIdx(),
+            separators.first(),
+            separators.second(),
+            separators.count(),
+            separators.lastSeparator());
+  }
+
+  /** Internal record for returning when a new format is matched. */
+  record NumberParseResultWithFormat(NumberWithSeparators format, NumberParseResult result)
+      implements NumberParseResult {}
+
+  /** Internal record for returning when a new format is matched. */
+  record NumberParseResultWithIndex(int endIdx, NumberParseResult result)
+      implements NumberParseResult {}
```

Comment:
Isn't the doc copied over from above?

What is the purpose of returning this index here? Why the `WithFormat` one does not need it?